### PR TITLE
SFX 253 - Add Products to SAYT

### PR DESCRIPTION
**Pull request by ChantalDesrochers**
- add products to sayt component
- added stories and knobs to stories to demonstrate products integration into sayt
- moved shared Storybook resources into storybook/common file
- removed storybook/common import from sayt/test/utitls to fix custom element registry error
NOTE - editor made formatting changes on merge/save. Updated most diffs to match Master (spacing, spaces, etc.). Confirmed w/ @DanielMoniz-GBI that minor diffs will be resolved once the linter is in place.

### CODE REVIEW

File: packages/web-components/@sfx/search-box/test/unit/search-box.test.ts [hunk 1]
```diff
@@ -95,7 +95,7 @@ describe('SearchBox Component', () => {
       const detail = 'inputText';
       const inputEvent = new CustomEvent('some-test-type', { detail });
       const updateSearchTermValue = stub(searchbox, 'updateSearchTermValue');
-  
+
```
Comment (jmbtutor) [inline packages/web-components/@sfx/search-box/test/unit/search-box.test.ts]: It may have been your formatter doing this, but since this is an unrelated file, please revert this change.

File: packages/web-components/@sfx/sayt/src/sayt.ts [hunk 2]
```diff
@@ -186,13 +190,32 @@ export default class Sayt extends LitElement {
...
-      }
-      ${ this.hideAutocomplete ? '' : html`<sfx-autocomplete></sfx-autocomplete>` }
+      <style>
+        .sfx-sayt-container {
+          display: flex;
+        }
+        sfx-autocomplete {
+          flex: 1;
+        }
+        sfx-products {
```
[the remark this answers is not in the document]
Comment (DanielMoniz-GBI) [reply packages/web-components/@sfx/sayt/src/sayt.ts]: Below you'll see that as part of the render method, there is a div with that class. The idea appears to be that the Close button should be within the overall SAYT wrapper, but not within the area being styled as `flex`. An additional wrapper is therefore needed. The client does not have to supply a class.

File: packages/web-components/@sfx/sayt/src/sayt.ts [hunk 3]
```diff
@@ -186,13 +190,32 @@ export default class Sayt extends LitElement {
...
+          flex: 2;
```
[the remark this answers is not in the document]
Comment (ChantalDesrochers) [reply packages/web-components/@sfx/sayt/src/sayt.ts]: The declarations are to determine how much space each component should get within the containing element. The declarations impact the 'size', not the order.
If I remove it, autocomplete only takes as much space as it needs, and it ends up looking squished on the left.

File: packages/web-components/@sfx/sayt/stories/index.ts [hunk 4]
```diff
@@ -1,29 +1,55 @@
 import { storiesOf } from '@storybook/html';
 import { withKnobs, text, boolean } from '@storybook/addon-knobs';
-import { getDisplayCode } from '../test/utils';
-import '../src/index';
+import '../src';
+import { getDisplayCode, sendSampleProducts, getProducts } from '../../../../../.storybook/common';
```
[the remark this answers is not in the document]
Comment (ChantalDesrochers) [reply packages/web-components/@sfx/sayt/stories/index.ts]: The file was causing the customelementregistry this name has already been used with this registry.
The removal of the utils file eliminated the problem. @jmbtutor is aware of the issue/solution.

File: packages/web-components/@sfx/sayt/stories/index.ts [hunk 5]
```diff
@@ -1,29 +1,55 @@
...
-  return '<sfx-sayt\n'
-    + (searchbox ? `  searchbox="${searchbox}"\n` : '')
-    + `  closetext="${closeText}"\n`
-    + (showCloseButton ? `  ${showCloseButton}\n` : '')
-    + (showAttribute ? `  ${showAttribute}\n` : '')
-    + '></sfx-sayt>';
+  const hideAutocomplete = boolean('Hide Autocomplete', false) ? 'hideAutocomplete' : '';
+  const hideProducts = boolean('Hide Products', false) ? 'hideProducts' : '';
+
+  return (
```
[the remark this answers is not in the document]
Comment (ChantalDesrochers) [reply packages/web-components/@sfx/sayt/stories/index.ts]: Those were added by my code formatter!

File: .storybook/common.ts [hunk 6]
```diff
@@ -1,16 +1,62 @@
...
         }
       </style>
       <h3>The code</h3>
-      <pre class="code-output"><code>${ entities.encode(code) }</code></pre>
-    `
+      <pre class="code-output"><code>${entities.encode(code)}</code></pre>
+    `;
+}
+
+export const sampleProducts = [
```
[the remark this answers is not in the document]
Comment (ChantalDesrochers) [reply .storybook/common.ts]: Done!

File: .storybook/common.ts [hunk 7]
```diff
@@ -1,16 +1,62 @@
...
+    price: 49.99,
+    label: 'Classic Product',
+    promo: '25% off',
+    imageSrc:
+      'https://images.unsplash.com/photo-1542291026-7eec264c27ff?ixlib=rb-1.2.1&auto=format&fit=crop&h=350&q=80',
+    imageAlt: 'A classic red shoe'
+  } as ProductModel
+];
+
+export function getProducts(quantity) {
```
[the remark this answers is not in the document]
Comment (ChantalDesrochers) [reply .storybook/common.ts]: Done!

File: packages/web-components/@sfx/sayt/CHANGELOG.md [hunk 8]
```diff
@@ -8,3 +8,4 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - SFX-191: Added the `sayt` component.
   - This component renders `sayt` related components such as `sfx-autocomplete` and `sfx-products`.
+- SFX-253: Added `sfx-products` to the autocomplete component.
```
[the remark this answers is not in the document]
Comment (ChantalDesrochers) [reply packages/web-components/@sfx/sayt/CHANGELOG.md]: Done!

File: packages/web-components/@sfx/sayt/stories/index.ts [hunk 9]
```diff
@@ -1,29 +1,55 @@
...
+    { title: 'Brands', items: [{ label: 'Cats' }, { label: 'Dogs' }] },
+    { title: '', items: [{ label: 'Cars' }, { label: 'Bikes' }] }
+  ],
+  bubbles: true
+});
+
+function getStyle() {
+  return `
+  <style>
+  * {
```
[the remark this answers is not in the document]
Comment (ChantalDesrochers) [reply packages/web-components/@sfx/sayt/stories/index.ts]: Done!

File: packages/web-components/@sfx/sayt/stories/index.ts [hunk 10]
```diff
@@ -35,74 +61,105 @@ function emitEventInFuture(event, timeout = 100) {
...
+      emitEventInFuture(autocompleteDataReceivedEvent, 500);
+      setTimeout(() => {
+        const products = getProducts(10);
+        sendSampleProducts(products);
+      }, 500);
+
+      const sayt = getSayt();
+      const style = getStyle();
+      return `
+      ${ style }
```
[the remark this answers is not in the document]
Comment (ChantalDesrochers) [reply packages/web-components/@sfx/sayt/stories/index.ts]: Done

File: packages/web-components/@sfx/sayt/stories/index.ts [hunk 11]
```diff
@@ -114,24 +171,82 @@ ${ sayt1 }
...
+    },
+    {
+      notes: {
+        markdown: `
         #Search As You Type (SAYT)
-        Demonstrating multiple SAYT components.This proves that each Search/SAYT
-        pair acts independently.
+        Demonstrating functionality of SAYT when products and autocomplete events are received at different points.
+        Each portion of SAYT (autocomplete, products) should display when its respective data is received.
+        Autocomplete should appear at 1 second, products should appear at 3 seconds.
```
[the remark this answers is not in the document]
Comment (ChantalDesrochers) [reply packages/web-components/@sfx/sayt/stories/index.ts]: Done

File: .storybook/common.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ChantalDesrochers) [reply .storybook/common.ts]: Done

File: packages/web-components/@sfx/sayt/src/sayt.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChantalDesrochers) [reply packages/web-components/@sfx/sayt/src/sayt.ts]: Removed the flex property from sfx-autocomplete and sfx-products completely -  after discussion with Daniel, this type of styling can be handled by the style focused tickets.

File: packages/web-components/@sfx/sayt/src/sayt.ts [hunk 12]
```diff
@@ -186,13 +190,32 @@ export default class Sayt extends LitElement {
...
     return html`
-      ${ this.showCloseButton ?
-        html`<button aria-label="Close" @click=${ this.clickCloseSayt }>
-          ${ this.closeText }
-        </button>`
-        : ''
-      }
-      ${ this.hideAutocomplete ? '' : html`<sfx-autocomplete></sfx-autocomplete>` }
+      <style>
+        .sfx-sayt-container {
```
[the remark this answers is not in the document]
Comment (ChantalDesrochers) [reply packages/web-components/@sfx/sayt/src/sayt.ts]: This additional div was added in order to style sfx-products and sfx-autocomplete with flex, but not the button.

File: .storybook/common.ts [hunk 13]
```diff
@@ -1,16 +1,71 @@
...
+    bubbles: true
+  });
+}
+
+export function sendSampleProducts(products: ProductModel[]) {
+  const productsEvent = getProductsReceivedEvent(products);
+  window.dispatchEvent(productsEvent);
+}
+
+// TODO update function name once events repo is integrated - 'PRODUCTS_EVENT' will have been renamed
```
Comment (jmbtutor) [inline .storybook/common.ts]: This comment belongs above `getProductsReceivedEvent`. You can also just leave it off because we'll see that it's broken once the events have been stripped out.

File: packages/web-components/@sfx/sayt/CHANGELOG.md [hunk 14]
```diff
@@ -8,3 +8,4 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - SFX-191: Added the `sayt` component.
   - This component renders `sayt` related components such as `sfx-autocomplete` and `sfx-products`.
+- SFX-253: Added `sfx-products` to dispay autocomplete products.
```
Comment (jmbtutor) [inline packages/web-components/@sfx/sayt/CHANGELOG.md]: I thought I had flagged this before, but since this is still unreleased, we can add this as a sub-point, or even not at all because it's covered by line 10.

Comment (ChantalDesrochers) [reply packages/web-components/@sfx/sayt/CHANGELOG.md]: Removing since Daniel also suggested it be removed.

File: packages/web-components/@sfx/sayt/src/sayt.ts [hunk 15]
```diff
@@ -194,14 +201,26 @@ export default class Sayt extends LitElement {
...
-      }
-      ${ this.hideAutocomplete ? '' : html`<sfx-autocomplete></sfx-autocomplete>` }
+        ${this.showCloseButton
+          ? html`
+              <button aria-label="Close" @click=${this.clickCloseSayt}>
+                ${this.closeText}
+              </button>
+            `
+          : ''}
+        <div class="sfx-sayt-container">
```
Comment (geoffcoutts) [inline packages/web-components/@sfx/sayt/src/sayt.ts]: Is this container necessary?

Comment (jmbtutor) [reply packages/web-components/@sfx/sayt/src/sayt.ts]: If we have a close button, I think it makes it easier to do flexbox if we have it.

File: .storybook/common.ts [hunk 16]
```diff
@@ -1,16 +1,70 @@
...
+    bubbles: true
+  });
+}
+
+export function sendSampleProducts(products: ProductModel[]) {
+  const productsEvent = getProductsReceivedEvent(products);
+  window.dispatchEvent(productsEvent);
+}
+
+export function dispatchProvideProductsEvent() {
```
[the remark this answers is not in the document]
Comment (jmbtutor) [reply .storybook/common.ts]: That makes sense. I'll add that.
